### ISSUE TITLE
Crash loop using Vault Agent Template

### ISSUE THREAD

**Issue by asaintsever** [hashicorp/vault#7791]
**Describe the bug**
I previously used Vault Agent+Consul Template to connect and fetch secrets from a Vault server. I am trying to leverage new Vault Agent Template feature (beta) to only have to deal with one sidecar container and achieve same result.
Using basic configuration and template that worked with Consul Template I notice infinite stop and restart of Vault Agent container due to a crash in the new template server (see provided logs below).

**To Reproduce**
Pre-requisites: having a running Vault server with some secrets under "secret/test/test-app-svc".

1. Deploy pod with Vault 1.3.0-beta1 image in agent mode with configuration as provided below (see Additional context section)
2. Tail Vault Agent container logs. The pod will run ok for a short while then will fail then restart and so on.

Error in Vault Agent container log:

```log
==> Vault server started! Log data will stream in below:

==> Vault agent configuration:

                     Cgo: disabled
               Log Level: info
                 Version: Vault v1.3.0-beta1

2019-11-05T14:52:24.783Z [INFO]  sink.file: creating file sink
2019-11-05T14:52:24.783Z [INFO]  sink.file: file sink configured: path=/home/vault/.vault-token mode=-rw-r-----
2019-11-05T14:52:24.784Z [INFO]  auth.handler: starting auth handler
2019-11-05T14:52:24.784Z [INFO]  auth.handler: authenticating
2019-11-05T14:52:24.785Z [INFO]  template.server: starting template server
2019-11-05 14:52:24.785856 I | [INFO] (runner) creating new runner (dry: false, once: false)
2019-11-05 14:52:24.786982 I | [DEBUG] (runner) final config: {"Consul":{"Address":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000}},"Dedup":{"Enabled":false,"MaxStale":2000000000,"Prefix":"consul-template/dedup/","TTL":15000000000},"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"WARN","MaxStale":2000000000,"PidFile":"","ReloadSignal":1,"Syslog":{"Enabled":false,"Facility":"LOCAL0"},"Templates":[{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"  {{ with secret \"secret/test/test-app-svc\" }}{{ range $k, $v := .Data }}\n{{ $k }}={{ $v }}\n{{ end }}{{ end }}\n\n  ","CreateDestDirs":true,"Destination":"/opt/test/secrets/secrets.properties","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"","Wait":{"Enabled":true,"Min":1000000000,"Max":2000000000},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""}],"Vault":{"Address":"http://vault:8200","Enabled":true,"Grace":15000000000,"Namespace":"","RenewToken":false,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":false},"Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false},"Wait":{"Enabled":false,"Min":0,"Max":0},"Once":false}
2019-11-05 14:52:24.787177 I | [INFO] (runner) creating watcher
2019-11-05T14:52:24.881Z [INFO]  sink.server: starting sink server
2019-11-05T14:52:24.912Z [INFO]  auth.handler: authentication successful, sending token to sinks
2019-11-05T14:52:24.913Z [INFO]  auth.handler: starting renewal process
2019-11-05T14:52:24.981Z [INFO]  sink.file: token written: path=/home/vault/.vault-token
2019-11-05T14:52:24.983Z [INFO]  template.server: template server received new token
2019-11-05 14:52:24.983790 I | [INFO] (runner) stopping
2019-11-05 14:52:24.984208 I | [DEBUG] (runner) stopping watcher
2019-11-05 14:52:24.984638 I | [DEBUG] (watcher) stopping all views
2019-11-05 14:52:24.985051 I | [INFO] (runner) creating new runner (dry: false, once: false)
2019-11-05 14:52:24.985672 I | [DEBUG] (runner) final config: {"Consul":{"Address":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000}},"Dedup":{"Enabled":false,"MaxStale":2000000000,"Prefix":"consul-template/dedup/","TTL":15000000000},"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"WARN","MaxStale":2000000000,"PidFile":"","ReloadSignal":1,"Syslog":{"Enabled":false,"Facility":"LOCAL0"},"Templates":[{"Backup":false,"Command":"","CommandTimeout":30000000000,"Contents":"  {{ with secret \"secret/test/test-app-svc\" }}{{ range $k, $v := .Data }}\n{{ $k }}={{ $v }}\n{{ end }}{{ end }}\n\n  ","CreateDestDirs":true,"Destination":"/opt/test/secrets/secrets.properties","ErrMissingKey":false,"Exec":{"Command":"","Enabled":false,"Env":{"Blacklist":[],"Custom":[],"Pristine":false,"Whitelist":[]},"KillSignal":2,"KillTimeout":30000000000,"ReloadSignal":null,"Splay":0,"Timeout":30000000000},"Perms":0,"Source":"","Wait":{"Enabled":true,"Min":1000000000,"Max":2000000000},"LeftDelim":"","RightDelim":"","FunctionBlacklist":null,"SandboxPath":""}],"Vault":{"Address":"http://vault:8200","Enabled":true,"Grace":15000000000,"Namespace":"","RenewToken":false,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":false},"Transport":{"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":3,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false},"Wait":{"Enabled":false,"Min":0,"Max":0},"Once":false}
2019-11-05 14:52:24.986277 I | [INFO] (runner) creating watcher
2019-11-05 14:52:24.986812 I | [INFO] (runner) starting
2019-11-05 14:52:24.987142 I | [DEBUG] (runner) running initial templates
2019-11-05 14:52:24.987375 I | [DEBUG] (runner) initiating run
2019-11-05 14:52:24.987745 I | [DEBUG] (runner) checking template 8329ee484254684062611b314191de5a
2019-11-05 14:52:24.988636 I | [DEBUG] (runner) was not watching 1 dependencies
2019-11-05 14:52:24.988696 I | [DEBUG] (watcher) adding vault.read(secret/test/test-app-svc)
2019-11-05 14:52:24.988713 I | [TRACE] (watcher) vault.read(secret/test/test-app-svc) starting
2019-11-05 14:52:24.988734 I | [DEBUG] (runner) diffing and updating dependencies
2019-11-05 14:52:24.988747 I | [DEBUG] (runner) watching 1 dependencies
2019-11-05 14:52:24.988828 I | [TRACE] (view) vault.read(secret/test/test-app-svc) starting fetch
2019-11-05T14:52:25.087Z [INFO]  auth.handler: renewed auth token
2019-11-05 14:52:25.092409 I | [TRACE] vault.read(secret/test/test-app-svc): GET /v1/secret/test/test-app-svc
2019-11-05 14:52:25.288890 I | [TRACE] vault.read(secret/test/test-app-svc): non-renewable secret, set sleep for 8.847714278s
2019-11-05 14:52:25.288978 I | [TRACE] (view) vault.read(secret/test/test-app-svc) marking successful data response
2019-11-05 14:52:25.289007 I | [TRACE] (view) vault.read(secret/test/test-app-svc) successful contact, resetting retries
2019-11-05 14:52:25.289023 I | [TRACE] (view) vault.read(secret/test/test-app-svc) received data
2019-11-05 14:52:25.289163 I | [TRACE] (view) vault.read(secret/test/test-app-svc) starting fetch
2019-11-05 14:52:25.289281 I | [DEBUG] (runner) receiving dependency vault.read(secret/test/test-app-svc)
2019-11-05 14:52:25.289302 I | [DEBUG] (runner) initiating run
2019-11-05 14:52:25.289314 I | [DEBUG] (runner) checking template 8329ee484254684062611b314191de5a
2019-11-05 14:52:25.289930 I | [DEBUG] (runner) rendering "(dynamic)" => "/opt/test/secrets/secrets.properties"
2019-11-05 14:52:25.290209 I | [DEBUG] (runner) diffing and updating dependencies
2019-11-05 14:52:25.290340 I | [DEBUG] (runner) vault.read(secret/test/test-app-svc) is still needed
2019-11-05 14:52:25.290358 I | [DEBUG] (runner) watching 1 dependencies
2019-11-05 14:52:25.290514 I | [DEBUG] (runner) all templates rendered
2019-11-05 14:52:25.290677 I | [DEBUG] (runner) enabling template-specific quiescence for "8329ee484254684062611b314191de5a"
2019-11-05 14:52:34.137527 I | [TRACE] vault.read(secret/test/test-app-svc): GET /v1/secret/test/test-app-svc
2019-11-05 14:52:34.192389 I | [TRACE] vault.read(secret/test/test-app-svc): non-renewable secret, set sleep for 9.207199272s
2019-11-05 14:52:34.281378 I | [TRACE] (view) vault.read(secret/test/test-app-svc) marking successful data response
2019-11-05 14:52:34.281956 I | [TRACE] (view) vault.read(secret/test/test-app-svc) successful contact, resetting retries
2019-11-05 14:52:34.282379 I | [TRACE] (view) vault.read(secret/test/test-app-svc) received data
2019-11-05 14:52:34.282975 I | [DEBUG] (runner) receiving dependency vault.read(secret/test/test-app-svc)
2019-11-05 14:52:34.283113 I | [DEBUG] (runner) initiating run
2019-11-05 14:52:34.283441 I | [TRACE] (view) vault.read(secret/test/test-app-svc) starting fetch
2019-11-05 14:52:34.284797 I | [DEBUG] (runner) checking template 8329ee484254684062611b314191de5a
2019-11-05 14:52:34.286390 I | [DEBUG] (runner) diffing and updating dependencies
2019-11-05 14:52:34.286877 I | [DEBUG] (runner) vault.read(secret/test/test-app-svc) is still needed
2019-11-05 14:52:34.287234 I | [DEBUG] (runner) watching 1 dependencies
2019-11-05 14:52:34.287775 I | [DEBUG] (runner) all templates rendered
2019-11-05 14:52:35.288445 I | [DEBUG] (runner) received template "8329ee484254684062611b314191de5a" from quiescence
2019-11-05 14:52:35.288963 I | [DEBUG] (runner) initiating run
2019-11-05 14:52:35.289292 I | [DEBUG] (runner) checking template 8329ee484254684062611b314191de5a
2019-11-05 14:52:35.290367 I | [DEBUG] (runner) rendering "(dynamic)" => "/opt/test/secrets/secrets.properties"
2019-11-05 14:52:35.381300 I | [DEBUG] (runner) diffing and updating dependencies
2019-11-05 14:52:35.381320 I | [DEBUG] (runner) vault.read(secret/test/test-app-svc) is still needed
2019-11-05 14:52:35.381324 I | [DEBUG] (runner) watching 1 dependencies
2019-11-05 14:52:35.381328 I | [DEBUG] (runner) all templates rendered
2019-11-05 14:52:35.381332 I | [DEBUG] (runner) enabling template-specific quiescence for "8329ee484254684062611b314191de5a"
2019-11-05T14:52:35.381Z [INFO]  template.server: template server stopped
panic: close of closed channel

goroutine 36 [running]:
github.com/hashicorp/vault/command/agent/template.(*Server).Run(0xc0006e6690, 0x3a9c520, 0xc0006e2b00, 0xc0006ec4e0, 0xc00000f020, 0x1, 0x1)
	/go/src/github.com/hashicorp/vault/command/agent/template/template.go:141 +0x762
created by github.com/hashicorp/vault/command.(*AgentCommand).Run
	/go/src/github.com/hashicorp/vault/command/agent.go:538 +0x210f

```

**Expected behavior**
Vault Agent should continue executing the template without errors and pod failure.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.2.2
* Vault CLI Version (retrieve with `vault version`): 1.2.2 (1.3.0-beta1 in deployed pod)
* Server Operating System/Architecture: Minikube 1.3.1 on Ubuntu 18.04.3 LTS, Kubernetes 1.15.2

Vault server configuration file(s):
Deployed in dev mode using latest official Helm chart (https://github.com/hashicorp/vault-helm).

**Additional context**
Vault Agent configuration:

```hcl
pid_file = "/home/vault/pidfile"

auto_auth {
  method "kubernetes" {
    mount_path = "auth/kubernetes"
    config = {
      role = "test"
    }
  }

  sink "file" {
    config = {
      path = "/home/vault/.vault-token"
    }
  }
}

template {
  destination = "/opt/test/secrets/secrets.properties"
  contents = <<EOH
  {{ with secret "secret/test/test-app-svc" }}{{ range $k, $v := .Data }}
{{ $k }}={{ $v }}
{{ end }}{{ end }}

  EOH
  command = ""
  wait {
    min = "1s"
    max = "2s"
  }
}
```

**Comment by catsby** [comment]
Hello - thank you for the detailed write up! I was able to reproduce the issue you described with the beta. I then ran the setup against a complied version of vault that included the changes and fixes to Agent Templating in https://github.com/hashicorp/vault/pull/7739 and confirmed that I could no longer reproduce the crash in that version. 

I'll leave this open for now, but assume this to be fixed when that PR is merged. 

Thanks!
